File: typescript/tests/predictive_engine.test.ts
// predictive_engine.test.ts
//
// Verifies the two IPredictiveEngine implementations against the C# reference:
//   HistogramPredictiveEngine (HerJarvisRealImplementations.cs #14) and
//   SequencePredictiveEngine  (SequencePredictiveEngine.cs).
//
// Both engines read the wall clock (DateTimeOffset.UtcNow) for "now", so the
// tests are written to be clock-independent: histogram tests anchor observed
// events to the CURRENT slot (which the m=0 horizon sample always hits), and
// sequence tests assert count-derived probabilities exactly while allowing a
// tolerance on the forecast timestamp.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  HistogramPredictiveEngine,
  SequencePredictiveEngine,
} from '../src/companion/reasoning/index';

describe('HistogramPredictiveEngine', () => {
  it('anticipates a need whose slot falls inside the horizon (prob = share)', async () => {
    const eng = new HistogramPredictiveEngine();
    const now = new Date();
    // Two hits at the current slot, one at a far slot (3.5 days out).
    eng.observe('coffee', now);
    eng.observe('coffee', now);
    eng.observe('coffee', new Date(now.getTime() + 84 * 3600 * 1000)); // ~3.5 days
    // Sub-30-min horizon → the sampling loop runs only m=0 (the current slot).
    // A 60-min horizon samples m=0 AND m=30, which land in the SAME hour slot
    // whenever the wall-clock minute is < 30, double-counting the current slot
    // (2→4) and making the C#-faithful result clock-minute-dependent (2/3 when
    // minute≥30, 4/3 when minute<30). The engine matches C# exactly; the
    // reference algorithm carries that latent clock-dependence, so the test
    // pins the deterministic sub-30 case.
    const needs = await eng.anticipateAsync(20); // only m=0 samples the current slot
    const coffee = needs.find((n) => n.description === 'coffee');
    assert.ok(coffee, 'coffee should be anticipated');
    // total = 3; upcoming = the 2 current-slot hits (far one not sampled) → 2/3.
    assert.equal(coffee!.probability, 2 / 3);
  });

  it('skips a need with zero occurrences inside the horizon', async () => {
    const eng = new HistogramPredictiveEngine();
    const now = new Date();
    // Put the only occurrence ~3.5 days out, far from any 30-min sample of a
    // 60-min horizon → not anticipated.
    eng.observe('gym', new Date(now.getTime() + 84 * 3600 * 1000));
    const needs = await eng.anticipateAsync(60);
    assert.ok(!needs.some((n) => n.description === 'gym'));
  });

  it('preserves the original description surface form (case-insensitive key)', async () => {
    const eng = new HistogramPredictiveEngine();
    const now = new Date();
    eng.observe('Coffee', now);
    eng.observe('coffee', now); // merges onto the first surface form
    const needs = await eng.anticipateAsync(60);
    const hit = needs.find((n) => n.description.toLowerCase() === 'coffee');
    assert.ok(hit);
    assert.equal(hit!.description, 'Coffee'); // first-seen form wins
  });

  it('orders results by descending probability', async () => {
    const eng = new HistogramPredictiveEngine();
    const now = new Date();
    // 'a' present once at current slot (prob 1). 'b' present at current slot
    // once plus far once (prob 0.5). So a should sort before b.
    eng.observe('a', now);
    eng.observe('b', now);
    eng.observe('b', new Date(now.getTime() + 84 * 3600 * 1000));
    const needs = await eng.anticipateAsync(60);
    const probs = needs.map((n) => n.probability);
    for (let i = 1; i < probs.length; i++) assert.ok(probs[i - 1] >= probs[i]);
    assert.equal(needs[0].description, 'a');
  });

  it('rejects a non-positive horizon and a blank description', async () => {
    const eng = new HistogramPredictiveEngine();
    await assert.rejects(() => eng.anticipateAsync(0), /horizonMinutes out of range/);
    assert.throws(() => eng.observe('  ', new Date()), /description required/);
  });
});

describe('SequencePredictiveEngine', () => {
  it('predicts the next event from the n-gram context with back-off weighting', async () => {
    const eng = new SequencePredictiveEngine(3);
    const base = new Date('2026-01-01T00:00:00Z').getTime();
    // Deterministic sequence: wake -> coffee -> email, wake -> coffee -> email.
    // After the second "coffee", the context {wake,coffee} strongly predicts
    // "email".
    const seq = ['wake', 'coffee', 'email', 'wake', 'coffee'];
    seq.forEach((e, i) => eng.observe(e, new Date(base + i * 3600 * 1000)));
    const needs = await eng.anticipateAsync(24 * 60);
    assert.ok(needs.length > 0);
    assert.equal(needs[0].description, 'email');
    // Probabilities are a normalised distribution → sum to ~1 over the emitted
    // (in-horizon) events; each is in (0,1].
    for (const n of needs) assert.ok(n.probability > 0 && n.probability <= 1);
  });

  it('forecasts arrival from the mean inter-arrival interval', async () => {
    const eng = new SequencePredictiveEngine(3);
    const base = new Date('2026-01-01T00:00:00Z').getTime();
    // Same event repeated at a fixed 10-minute cadence → mean interval 600s.
    // (Inter-arrival is only tracked when the immediately preceding event is
    // the same event, so a pure repeat sequence exercises it.)
    for (let i = 0; i < 4; i++) eng.observe('ping', new Date(base + i * 10 * 60 * 1000));
    const before = Date.now();
    const needs = await eng.anticipateAsync(60); // horizon 3600s > 600s interval
    const after = Date.now();
    const ping = needs.find((n) => n.description === 'ping');
    assert.ok(ping, 'ping should be anticipated within the horizon');
    // expectedByUtc ≈ now + 600s. Allow the wall-clock window between our two
    // Date.now() reads plus a small slack.
    const expectedLo = before + 600 * 1000 - 50;
    const expectedHi = after + 600 * 1000 + 50;
    const got = ping!.expectedByUtc.getTime();
    assert.ok(got >= expectedLo && got <= expectedHi, `forecast ${got} not in [${expectedLo},${expectedHi}]`);
  });

  it('drops events whose mean interval exceeds the horizon', async () => {
    const eng = new SequencePredictiveEngine(3);
    const base = new Date('2026-01-01T00:00:00Z').getTime();
    // 'slow' repeats every 2 hours (7200s) → exceeds a 60-min (3600s) horizon.
    for (let i = 0; i < 3; i++) eng.observe('slow', new Date(base + i * 2 * 3600 * 1000));
    const needs = await eng.anticipateAsync(60);
    assert.ok(!needs.some((n) => n.description === 'slow'));
  });

  it('returns [] before any events are observed', async () => {
    const eng = new SequencePredictiveEngine();
    assert.deepEqual(await eng.anticipateAsync(60), []);
  });

  it('rejects an out-of-range order and a non-positive horizon', async () => {
    assert.throws(() => new SequencePredictiveEngine(0), /order out of range/);
    assert.throws(() => new SequencePredictiveEngine(7), /order out of range/);
    const eng = new SequencePredictiveEngine();
    await assert.rejects(() => eng.anticipateAsync(-5), /horizonMinutes out of range/);
    assert.throws(() => eng.observe('', new Date()), /event required/);
  });
});

File: rust/tests/predictive_engine_test.rs
//! predictive_engine_test.rs
//!
//! Verifies the companion predictive engine: HistogramPredictiveEngine
//! (day-of-week × hour histogram) and SequencePredictiveEngine (variable-order
//! Markov chain with back-off + inter-arrival forecasting). Mirrors the C#
//! IPredictiveEngine.AnticipateAsync behaviour.

use chrono::{Duration, Utc};
use circle_ai::companion::predictive_engine::{
    HistogramPredictiveEngine, IPredictiveEngine, SequencePredictiveEngine,
};

// ── HistogramPredictiveEngine ──────────────────────────────────────────────

#[test]
fn histogram_empty_engine_returns_nothing() {
    let e = HistogramPredictiveEngine::new();
    assert!(e.anticipate(120).is_empty());
}

#[test]
fn histogram_anticipates_an_event_seen_at_the_current_slot() {
    let e = HistogramPredictiveEngine::new();
    // Observe at "now" so the m=0 sample of the horizon walk hits the same slot.
    e.observe("check-email", Utc::now());
    // Horizon < 30 ⇒ the walk samples only m=0, so exactly one slot is counted.
    // (The C# double-counts a slot when two 30-min samples land in the same
    // clock hour; a sub-30-min horizon avoids that so the probability is exact.)
    let needs = e.anticipate(1);
    assert_eq!(needs.len(), 1);
    assert_eq!(needs[0].description, "check-email");
    // Only one observation ⇒ upcoming/total == 1.
    assert!((needs[0].probability - 1.0).abs() < 1e-9);
}

#[test]
fn histogram_skips_events_outside_the_horizon_window() {
    let e = HistogramPredictiveEngine::new();
    // 12 hours from now is well outside a 60-minute horizon.
    e.observe("midnight-task", Utc::now() + Duration::hours(12));
    let needs = e.anticipate(60);
    assert!(
        needs.iter().all(|n| n.description != "midnight-task"),
        "far-off event must not be anticipated in a 60-min horizon"
    );
}

#[test]
fn histogram_orders_by_descending_probability() {
    let e = HistogramPredictiveEngine::new();
    let now = Utc::now();
    // "frequent" occurs entirely at the current slot (prob 1.0). "rare" occurs
    // once at the current slot and once 12h away (prob 0.5 within the window).
    e.observe("frequent", now);
    e.observe("rare", now);
    e.observe("rare", now + Duration::hours(12));
    let needs = e.anticipate(60);
    // Both present; frequent (1.0) must sort before rare (0.5).
    assert!(needs.len() >= 2);
    let freq_pos = needs.iter().position(|n| n.description == "frequent").unwrap();
    let rare_pos = needs.iter().position(|n| n.description == "rare").unwrap();
    assert!(freq_pos < rare_pos);
    assert!(needs[freq_pos].probability >= needs[rare_pos].probability);
}

#[test]
fn histogram_case_insensitive_description() {
    let e = HistogramPredictiveEngine::new();
    let now = Utc::now();
    e.observe("Standup", now);
    e.observe("standup", now);
    // Sub-30-min horizon ⇒ single sample, so the folded bucket reads prob 1.0.
    let needs = e.anticipate(1);
    // Both fold to one histogram bucket.
    assert_eq!(needs.len(), 1);
    assert!((needs[0].probability - 1.0).abs() < 1e-9);
}

#[test]
#[should_panic(expected = "horizonMinutes out of range")]
fn histogram_rejects_non_positive_horizon() {
    let e = HistogramPredictiveEngine::new();
    e.anticipate(0);
}

#[test]
#[should_panic(expected = "description required")]
fn histogram_observe_rejects_blank_description() {
    let e = HistogramPredictiveEngine::new();
    e.observe("  ", Utc::now());
}

// ── SequencePredictiveEngine ───────────────────────────────────────────────

#[test]
fn sequence_empty_engine_returns_nothing() {
    let e = SequencePredictiveEngine::default();
    assert!(e.anticipate(120).is_empty());
}

#[test]
fn sequence_predicts_the_learned_next_event() {
    let e = SequencePredictiveEngine::new(3);
    let base = Utc::now();
    // Teach the chain a -> b -> c a few times so the context [a,b,c...] leads to
    // a strong "wake -> coffee" style transition.
    for i in 0..4 {
        let t = base + Duration::minutes(i * 10);
        e.observe("wake", t);
        e.observe("coffee", t + Duration::minutes(1));
        e.observe("commute", t + Duration::minutes(2));
    }
    // A generous horizon so mean-interval defaults don't exclude candidates.
    let needs = e.anticipate(600);
    assert!(!needs.is_empty(), "expected at least one anticipated event");
    // Probabilities are normalised (sum over emitted candidates ≤ 1 by design of
    // the softmax-like aggregation; each is in (0,1]).
    for n in &needs {
        assert!(n.probability > 0.0 && n.probability <= 1.0);
    }
}

#[test]
fn sequence_drops_events_whose_mean_interval_exceeds_horizon() {
    let e = SequencePredictiveEngine::new(2);
    let base = Utc::now();
    // Same event repeated with a large (1 hour) gap ⇒ mean interval 3600s.
    e.observe("hourly", base);
    e.observe("hourly", base + Duration::hours(1));
    e.observe("hourly", base + Duration::hours(2));
    // Horizon of 10 minutes (600s) is far below the 3600s mean interval.
    let needs = e.anticipate(10);
    assert!(
        needs.iter().all(|n| n.description != "hourly"),
        "event with mean interval > horizon must be dropped"
    );
}

#[test]
fn sequence_is_case_sensitive() {
    let e = SequencePredictiveEngine::new(2);
    let base = Utc::now();
    // "A" and "a" are distinct contexts/events (Ordinal in the C#).
    e.observe("A", base);
    e.observe("b", base + Duration::minutes(1));
    e.observe("a", base + Duration::minutes(2));
    e.observe("c", base + Duration::minutes(3));
    // Context now ends in [a, c] — the [A,b]->? transition is a different key,
    // so no crash and predictions come only from lowercase-a's history.
    let _ = e.anticipate(600); // must not panic on the distinct-case keys
}

#[test]
#[should_panic(expected = "order out of range")]
fn sequence_rejects_order_zero() {
    let _ = SequencePredictiveEngine::new(0);
}

#[test]
#[should_panic(expected = "order out of range")]
fn sequence_rejects_order_above_six() {
    let _ = SequencePredictiveEngine::new(7);
}

#[test]
#[should_panic(expected = "event required")]
fn sequence_observe_rejects_blank_event() {
    let e = SequencePredictiveEngine::default();
    e.observe("", Utc::now());
}

#[test]
#[should_panic(expected = "horizonMinutes out of range")]
fn sequence_rejects_non_positive_horizon() {
    let e = SequencePredictiveEngine::default();
    e.anticipate(-5);
}
